code/game/machinery/doors/firedoor.dm

```diff
@@ -427,46 +447,83 @@
 /obj/machinery/door/firedoor/do_animate(animation)
 	switch(animation)
 		if("opening")
-			flick("door_opening", src)
+			flick("opening", src)
 			playsound(src, 'sound/machines/firelockopen.ogg', 37, 1)
 		if("closing")
 			playsound(src, 'sound/machines/firelockclose.ogg', 37, 1)
-			flick("door_closing", src)
+			flick("closing", src)
 	return
 
 
 /obj/machinery/door/firedoor/update_icon()
+	var/icon/lights_overlay
+
 	cut_overlays()
-	var/list/overlays_to_add = list()
+	set_light(0)
+	var/do_set_light = FALSE
 
 	if(density)
-		icon_state = "door_closed"
-		if(prying)
-			icon_state = "prying_closed"
-		if(hatch_open)
-			overlays_to_add += "hatch"
-		if(blocked)
-			overlays_to_add += "welded"
+		icon_state = "closed"
+		if(panel_open)
+			add_overlay(panel_file)
 		if(pdiff_alert)
-			overlays_to_add += "palert"
+			lights_overlay += "palert"
+			do_set_light = TRUE
 		if(dir_alerts)
 			for(var/d=1;d<=4;d++)
-				var/cdir = GLOB.cardinal[d]
 				for(var/i=1;i<=ALERT_STATES.len;i++)
-					if(dir_alerts[d] & (1<<(i-1)))
-						overlays_to_add += new/icon(icon,"alert_[ALERT_STATES[i]]", dir=cdir)
+					if(dir_alerts[d] & BITFLAG(i-1))
+						add_overlay("alert_[ALERT_STATES[i]]")
+						do_set_light = TRUE
 	else
-		icon_state = "door_open"
-		if(prying)
-			icon_state = "prying_open"
-		if(blocked)
-			overlays_to_add += "welded_open"
+		if(low_profile)
+			icon_state = "open_lowprofile"
+		else
+			icon_state = "open"
 
-	add_overlay(overlays_to_add)
+	if(blocked)
+		add_overlay(welded_file)
+
+	if(do_set_light)
+		set_light(2, 0.25, COLOR_SUN)
 
 	return
 
-//These are playing merry hell on ZAS.  Sorry fellas :(
+/obj/machinery/door/firedoor/setDir(ndir)
+	. = ..()
+	for(var/D in GLOB.cardinal)
+		var/turf/T = get_step(src, D)
+		for(var/obj/A in T.contents)
+			if(A.type in GLOB.firelock_align_types) //this is mainly for the cases where mappers can't manually align firelocks, i.e window spawners.
+				switch(D)
+					if(NORTH)
+						dir = WEST
+						break
+					if(EAST)
+						dir = SOUTH
+						break
+					if(SOUTH)
+						dir = WEST
+						break
+					if(WEST)
+						dir = SOUTH
+						break
+		if(T.density)
+			switch(D)
+				if(NORTH)
+					dir = WEST
+					break
+				if(EAST)
+					dir = SOUTH
+					break
+				if(SOUTH)
+					dir = WEST
+					break
+				if(WEST)
+					dir = SOUTH
+					break
```

Comment:
dir setting code should be before the . =..() and instead set ndir, so the proc itself sets it

setting ndir will make this happen because args are just passed down the ..() instead of copied so by re-setting an arg, it's propagating down.

directly dir setting after ..() will break things that rely on this behavior